python/device/device.py

```diff
@@ -78,16 +78,47 @@ class DeviceRunConfig(object):
     Currently hardcoded. Later, this will come from desired properties.
     """
 
+    # All durations are in seconds
     def __init__(self):
-        self.max_run_duration = 90
+        # how long should teh test run before finishing.  0 = forever
+        self.max_run_duration = 0
 
+        # How often do we update reported properties
         self.thief_property_update_interval = 60
+
+        # How long can a thread go without updating its watchdog before failing.
         self.watchdog_failure_interval = 300
 
-        self.d2c_operations_per_second = 5
-        self.d2c_slowness_threshold = 60
-        self.d2c_failures_allowed = 10
-        self.d2c_slow_operations_allowed = 100
+        # How long to keep trying to pair with a service instance before giving up.
+        self.pairing_request_timeout_interval = 900
+
+        # How many seconds to wait before sending a new pairingRequest message
+        self.pairing_request_send_interval = 30
+
+        # How many times to call send_message per second
+        self.send_message_operations_per_second = 1
+
+        # How many threads do we spin up for overlapped send_message calls
+        self.send_message_thread_count = 10
```

Comment:
Can you elaborate? Does this mean, if we're disconnected, we'll have, at max, 10 events waiting to send? That means if we disconnect longer than 10 seconds, we start to lose events that otherwise would have been sent?

Reply:
We have a queue of outgoing messages and 10 threads pulling out of the queue and sending.  We do this because send_message blocks until the PUBACK comes back from the service (about 20 ms).  If we have a single thread pulling off the queue and sending messages one-at-a-time, performance drops through the floor.  If the connection drops, we just let the queue grow (no max size) until some of these threads get freed up by arriving PUBACKs.

---
In reply to: [512844308](https://github.com/Azure/iot-sdk-longhaul/pull/17#discussion_r512844308) [](ancestors = 512844308)

Reply:
Ah, okay. Input queue, outgoing queue with 10 threads pulling off that queue. Sounds good.